### PR TITLE
added Slave IP and Hostname to remote state

### PR DESCRIPTION
**Pull request by ashvds**
Assumes a default network interface for AWS of "eth0", but allows for network interface to be specified in config.

### CODE REVIEW

**Comment by ashvds** [comment]
fixes #8

**Comment by ashvds** [comment]
Now also fixes #77 
status --show-slaves now looks like :
```
---------------------MASTER-------------------------------------------
Endpoints:
  main: (main.localhost:80) -> test1.zip
  test: (test.localhost:80) -> (not connected)

Deploys:
  test1.zip: (localhost:8000)
---------------------SLAVES-------------------------------------------
Slave: myslave
Slave IP: 192.168.0.75
Slave Hostname: laptop
Status: OK
Updated: 2019-09-11 03:55:15 UTC
Endpoints:
  main: (main.localhost:80) -> (not connected)
  test: (test.localhost:80) -> (not connected)

Deploys:
  test1.zip: (localhost:8000)
----------------------------------------------------------------------
Slave: myserver
Slave IP: 192.168.0.75
Slave Hostname: laptop
Status: OK
Updated: 2019-09-11 03:48:54 UTC
Endpoints:
  main: (main.localhost:80) -> (not connected)
  test: (test.localhost:80) -> (not connected)

Deploys:
  test1.zip: (localhost:8000)
----------------------------------------------------------------------

```

**Comment by ashvds** [comment]
This may be a breaking change for stale slave remote states.

**Review by timbod7 — COMMENTED** [review]
ListKeyPrefixes and listObjects are the same except for the lens accessor applied to each object returned. Hence it would be preferable to have one function, and map the accessor over the objects, ie something like:

```
listS3Objects  :: Env -> S3.BucketName -> S3.ObjectKey -> IO [S3.Object]
listS3Objects ...

listKeyPrefixes :: Env -> S3.BucketName -> S3.ObjectKey -> IO [T.Text]
listKeyPrefixes env bucketName objectKey
  =  (map (\o -> view S3.oKey o ^? S3.keyPrefix '/') <$> listS3Objects env bucketName objectKey
```

**Comment by ashvds** [comment]
@timbod7 , I added logging to S3 if someone specified a network interface that doesn't exist.
https://github.com/helix-collective/hx-deploy-tool/pull/75/commits/d502bbdb05dfafef711f67067a399804438a837e

This is the first haskell I felt I wrote instead of just repeating a pattern.
Love to hear your thoughts

File: adl/config.adl [hunk 1]
```diff
@@ -60,6 +60,11 @@ struct ProxyModeConfig {
     /// How we generate identifiers for slave machines
     MachineLabel slaveLabel = "ec2InstanceId";
 
+    // The network interface for retrieving the IP of the slave
+    // This will be logged to the remote state and allow easy identification and location of slave server if required
+    // In AWS interface defaults to eth0, this allows for other interfaces to be configured for ssh
+    String slaveSSHInterfaceName = "eth0";
```
Comment (timbod7) [inline adl/config.adl]: The interface doesn't have anything specific  to do with ssh. I'd jusdt call it `slaveInterfaceName`

File: adl/state.adl [hunk 2]
```diff
@@ -18,10 +18,16 @@ struct State {
...
 
 
-// The reported state of a slave
+// The (remotely) reported state of a slave
 struct SlaveState {
   // The result of the last state change
   SlaveStatus status;
+  
+  // The IP of the slave on which camus2 is running
+  String slaveIP;
```
Comment (timbod7) [inline adl/state.adl]: `slave` is redundant here. Everything in this struct is for the slave. Hence perhaps `ipAddress`,

File: adl/state.adl [hunk 3]
```diff
@@ -18,10 +18,16 @@ struct State {
...
+
+  //Host name of the slave
+  String slaveHostName;
```
Comment (timbod7) [inline adl/state.adl]: ditto.  Just `hostName`

File: src/Commands/ProxyMode/RemoteState.hs [hunk 4]
```diff
@@ -64,13 +64,13 @@ getSlaves remoteStateS3 = do
     getSlaveLabels :: Env -> IOR [T.Text]
     getSlaveLabels env = do
       let (bucketName,objectKey) = S3.splitPath (remoteStateS3 <> "/slaves")
-      keys <- liftIO $ S3.listObjects env bucketName objectKey
-      let labels = catMaybes (map parseSlaveLabel keys)
+      keysAndNames <- liftIO $ S3.listObjectsPrefixesAndNames env bucketName objectKey
+      let labels = catMaybes (map parseSlaveLabel (map fst keysAndNames))
```
Comment (timbod7) [inline src/Commands/ProxyMode/RemoteState.hs:68]: Perhaps leaving out the tuple would be clearer here, ie

```
objects <-  liftIO $ S3.listObjects env bucketName objectKey
let labels =  (map parseSlaveLabel  . catMaybes . map S3.s3Prefix) objects
```

File: adl/config.adl (hunk 1 above)
Comment (ashvds) [reply adl/config.adl]: Updated

File: src/Commands/ProxyMode/RemoteState.hs (hunk 4 above)
Comment (ashvds) [reply src/Commands/ProxyMode/RemoteState.hs:68]: agreed to leave as is

File: adl/state.adl (hunk 3 above)
Comment (ashvds) [reply adl/state.adl]: updated

**Review by paul-thompson-helix — APPROVED** [review]
LGTM on code walkthrough, merges and compiles ok.
Demo test still works..

**Comment by ashvds** [comment]
@paul-thompson-helix , concur, compiles, and running through the example still works. Merging!

**Pull request merged**